pkg/releaselock/rpc.go

```diff
@@ -0,0 +1,127 @@
+// clientIP is the internal IP address of the GKE node.
+func ReleaseLock(hostIP, clientIP string) error {
+	procedureID := sunrpc.ProcedureID{
+		ProgramNumber:   inbandLockReleaseProgramNumber,
+		ProgramVersion:  inbandLockReleaseProgramVersion,
+		ProcedureNumber: inbandLockReleaseProcedureNumber,
+	}
+	procedure := sunrpc.Procedure{
+		ID:   procedureID,
+		Name: inbandLockReleaseProcedureName,
+	}
+	// TODO(tyuchn): move the registration out to driver initialization.
+	if err := sunrpc.RegisterProcedure(procedure, true /* validateProcName */); err != nil {
+		return fmt.Errorf("failed to register procedure %+v: %w", procedure, err)
+	}
+
+	// Get port from portmapper.
+	hostAddress := fmt.Sprintf("%s:%s", hostIP, pmapPort)
+	port, err := sunrpc.PmapGetPort(hostAddress, pmapProgramNumber, pmapProgramVersion, sunrpc.IPProtoTCP)
+	if err != nil {
+		return fmt.Errorf("failed to get port for %s: %w", hostAddress, err)
+	}
+
+	// Connect to RPC server.
+	serverAddress := fmt.Sprintf("%s:%s", hostIP, strconv.Itoa(int(port)))
+	conn, err := net.Dial(protocol, serverAddress)
+	if err != nil {
+		return fmt.Errorf("failed to connect to RPC server at address %s: %w", serverAddress, err)
+	}
+
+	// Get notified when server closes the connection.
+	notifyClose := make(chan io.ReadWriteCloser)
+	go func() {
+		for rwc := range notifyClose {
+			conn := rwc.(net.Conn)
+			klog.Infof("Server %s disconnected", conn.RemoteAddr().String())
+		}
+	}()
+
+	// Create client using sunrpc codec.
+	client := sunrpc.NewClientCodec(conn, notifyClose)
+
+	klog.Infof("Calling Filestore IP %s to release all locks for GKE node %s", serverAddress, clientIP)
+
+	ip := net.ParseIP(clientIP)
+	if ip == nil {
+		return fmt.Errorf("invalid GKE node IP %s", clientIP)
+	}
+	ipByte := ip.To4()
+	if ipByte == nil {
+		return fmt.Errorf("invalid GKE node IPv4 address %s", clientIP)
+	}
+	ipBinary := binary.BigEndian.Uint32(ipByte)
+
+	request := rpc.Request{
+		ServiceMethod: inbandLockReleaseProcedureName,
+		Seq:           uint64(time.Now().UnixNano()),
+	}
+	if err := client.WriteRequest(&request, ipBinary); err != nil {
+		return fmt.Errorf("failed to write RPC request %+v for GKE node IP %s Filestore IP %s, err: %w", request, clientIP, hostIP, err)
+	}
+
+	response := rpc.Response{}
+	if err := client.ReadResponseHeader(&response); err != nil {
+		return fmt.Errorf("failed to read RPC response header for GKE node IP %s Filestore IP %s, err: %w", clientIP, hostIP, err)
+	}
+
+	var releaseAllLocksRes releaseLockResponse
+	if err := client.ReadResponseBody(&releaseAllLocksRes); err != nil {
+		return fmt.Errorf("failed to read RPC response body for GKE node IP %s Filestore IP %s, err: %w", clientIP, hostIP, err)
+	}
+	if releaseAllLocksRes.status != 0 {
+		return fmt.Errorf("failed to release all locks for GKE node IP %s Filestore IP %s, err: permission denied", clientIP, hostIP)
+	}
+
+	klog.Infof("Locks released for GKE node IP %s Filestore IP %s", serverAddress, clientIP)
```

Comment:
Please fix. node Ip is ClientIP and filestore IP is hostIP

Reply:
Done